python/dataproc_templates/gcs/gcs_to_spanner.py

```diff
@@ -0,0 +1,160 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict, Sequence, Optional, Any
+from logging import Logger
+import argparse
+import pprint
+
+from pyspark.sql import SparkSession, DataFrame
+
+from dataproc_templates import BaseTemplate
+import dataproc_templates.util.template_constants as constants
+
+__all__ = ['GCSToSpannerTemplate']
+
+
+class GCSToSpannerTemplate(BaseTemplate):
+  """
+  Dataproc template implementing loads from GCS into Spanner
+  """
+
+  @staticmethod
+  def parse_args(args: Optional[Sequence[str]] = None) -> Dict[str, Any]:
+    parser: argparse.ArgumentParser = argparse.ArgumentParser()
+
+    parser.add_argument(
+      f'--{constants.GCS_SPANNER_INPUT_LOCATION}',
+      dest=constants.GCS_SPANNER_INPUT_LOCATION,
+      required=True,
+      help='GCS location of the input files'
+    )
+    parser.add_argument(
+      f'--{constants.GCS_SPANNER_INPUT_FORMAT}',
+      dest=constants.GCS_SPANNER_INPUT_FORMAT,
+      required=True,
+      # TODO: Add support to CSV?
+      help='Input file format (one of: <avro,parquet,orc>)',
+      choices=[
+        constants.FORMAT_AVRO,
+        constants.FORMAT_PRQT,
+        constants.FORMAT_ORC,
+      ]
+    )
+    parser.add_argument(
+      f'--{constants.GCS_SPANNER_OUTPUT_INSTANCE}',
+      dest=constants.GCS_SPANNER_OUTPUT_INSTANCE,
+      required=True,
+      help='Spanner output instance'
+    )
+    parser.add_argument(
+      f'--{constants.GCS_SPANNER_OUTPUT_DATABASE}',
+      dest=constants.GCS_SPANNER_OUTPUT_DATABASE,
+      required=True,
+      help='Spanner output database'
+    )
+    parser.add_argument(
+      f'--{constants.GCS_SPANNER_OUTPUT_TABLE}',
+      dest=constants.GCS_SPANNER_OUTPUT_TABLE,
+      required=True,
+      help='Spanner output table'
+    )
+    parser.add_argument(
+      f'--{constants.GCS_SPANNER_PRIMARY_KEY}',
+      dest=constants.GCS_SPANNER_PRIMARY_KEY,
+      required=False,
+      help=(
+        "Primary key of the table. "
+        "It is only required when the table doesn't exist"
+      )
+    )
+    parser.add_argument(
+      f'--{constants.GCS_SPANNER_OUTPUT_SAVE_MODE}',
+      dest=constants.GCS_SPANNER_OUTPUT_SAVE_MODE,
+      required=False,
+      default=constants.OUTPUT_MODE_APPEND,
+      help=(
+        'Output write mode '
+        '(one of: <Overwrite|ErrorIfExists|Append|Ignore>) '
+        '(Defaults to ErrorIfExists)'
```

Comment:
The help and the docs in the README state the save mode defaults to `errorifexists`, but it actually defaults to `append`. Other Python templates default to `append`; so I'd suggest keeping that default, but updating the docs and help.

As a side note, I think defaulting to `errorifexists` is worth discussing. That's [Spark's built-in default](https://spark.apache.org/docs/latest/sql-data-sources-load-save-functions.html#save-modes) and we should probably respect that to be consistent with the ecosystem.

Reply:
You are right. I am changing the default to `errorifexists` for consistency (since this way is defined for the Java template)

Reply:
Sure! I'll see about creating an issue to move existing Python templates to this conventions. Resolving this thread.